### DISCUSSION TITLE
Specify the type of parameters separately instead of casting at every use site

### DISCUSSION THREAD

**Discussion by CodesInChaos** [geldata/gel#1580] [Ideas]
Currently when using parameters they need to be cast to the desired type at every use site. I consider this suboptimal for several reasons:

1. It adds noise to the query itself
2. Different uses can use different types, which makes it difficult to define a type for a parameter (e.g. it IDE hints or documentation)
3. In statically typed languages the type of the parameters in the query can be derived from the types defined in code

As an alternative I propose adding an optional separate section at the beginning of the query which assigns types to the parameters.

```
PARAMETERS {
    required data -> tuple<
        title: str,
        body: str,
        authors: array<tuple<
            name: str,
            email: str
        >>
    >;
}
INSERT Text {
    title := data.title,
    body := data.body,
    author := (
        FOR user IN array_unpack(data.authors)
        UNION (
            INSERT User {
                name := user.name,
                email := user.email,
             }
        )
    )
};
```

The driver could generate the `PARAMETERS` statement and combine it with the query using simple string concatenation.

**Comment by 1st1** [comment]
You can already do this:

```
WITH
  data := (
    title := <str>$title,
    body := <str>$body,
  )
}
INSERT Text {
    title := data.title,
    body := data.body
};
```

meaning that the proposed `PARAMETERS` is a syntax sugar. We'd like to avoid pushing non-essential changes to EdgeQL before we release 1.0 and start receiving more feedback from the community.

There's another limitation of current EdgeQL though: we cannot yet accept tuple query arguments. This is a limitation of Postgres, but I think we should be able to work around it. We'll basically need some new low-level mechanism to decode and re-encode tuples. Once that mechanism is implemented we can consider sugar syntax like `PARAMETERS`. In the meantime you can use JSON to pass tuples or arrays of tuples in.
